### PR TITLE
feat: use median timestamp

### PR DESCRIPTION
**Pull request by joske**
## Motivation
Use the median timestamp of the consensus output to have a deterministic timestamp

## Related PRs

https://github.com/AleoHQ/snarkOS/pull/2475

### CODE REVIEW

File: ledger/src/advance.rs [hunk 1]
```diff
@@ -22,6 +22,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         candidate_transactions: Vec<Transaction<N>>,
         candidate_solutions: Option<Vec<ProverSolution<N>>>,
         rng: &mut R,
+        median_timestamp: i64,
```
Comment (vvp) [inline ledger/src/advance.rs]: Perhaps rename this as `block_timestamp`, `next_timestamp` or just `timestamp`?  Talking about median timestamp here leaks unnecessary details of the current implementation.

File: ledger/src/tests.rs [hunk 2]
```diff
@@ -128,7 +129,15 @@ finalize foo:
...
 
     // Construct the next block.
-    let block = ledger.prepare_advance_to_next_block(&private_key, vec![transaction], None, rng).unwrap();
+    let block = ledger
+        .prepare_advance_to_next_block(
+            &private_key,
+            vec![transaction],
+            None,
+            rng,
+            OffsetDateTime::now_utc().unix_timestamp(),
```
Comment (vvp) [inline ledger/src/tests.rs]: Add assertion that the `block.timestamp()`  equals to the given timestamp.

**Review by vvp — CHANGES_REQUESTED** [review]
vn :+1:  left few comments

**Review by niklaslong — APPROVED** [review]

**Comment by niklaslong** [comment]
lgtm (pending Vesku's suggestions ofc)

**Review by vvp — APPROVED** [review]

**Review by ljedrz — APPROVED** [review]

File: ledger/src/advance.rs [hunk 3]
```diff
@@ -22,6 +22,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         candidate_transactions: Vec<Transaction<N>>,
         candidate_solutions: Option<Vec<ProverSolution<N>>>,
         rng: &mut R,
+        next_timestamp: i64,
```
Comment (howardwu) [inline ledger/src/advance.rs:25]: We should be passing in `candidate_certificates` to calculate the median timestamp here.

Let's get those data structures in place, then revisit this.

**Pull request closed without merging**